### ISSUE TITLE
cross-subnet does not work with route reflector

### ISSUE THREAD

**Issue by nightmeng** [projectcalico/calico#907]
I have deployed calico with route reflector on my test cluster, and I have enabled the IPinIP and cross-subnet.But the cross-subnet can only work when the route reflector and the node are in the same subnet. If the route reflector and the node are in different subnet, all the nodes communicate with each other over IPinIP.

```
[root@node-1 ~]# calicoctl node status
Calico process is running.

IPv4 BGP status
+---------------+---------------+-------+----------+-------------+
| PEER ADDRESS  |   PEER TYPE   | STATE |  SINCE   |    INFO     |
+---------------+---------------+-------+----------+-------------+
| 172.18.13.163 | node specific | up    | 11:04:39 | Established |
+---------------+---------------+-------+----------+-------------+

IPv6 BGP status
No IPv6 peers found.

[root@node-1 ~]# calicoctl get bgpPeer -o wide
SCOPE   PEERIP          NODE     ASN
node    172.18.13.163   node-1   64512
node    172.18.13.163   node-2   64512

[root@node-1 ~]# calicoctl get node -o wide
NAME     ASN       IPV4               IPV6
node-1   (64512)   172.18.10.145/24
node-2   (64512)   172.18.10.144/24

[root@node-1 ~]# route -n
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         172.18.10.254   0.0.0.0         UG    101    0        0 enp0s3
10.233.65.64    172.18.10.144   255.255.255.192 UG    0      0        0 tunl0
10.233.70.64    0.0.0.0         255.255.255.192 U     0      0        0 *
10.233.70.65    0.0.0.0         255.255.255.255 UH    0      0        0 calid82b120b044
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
172.18.10.0     0.0.0.0         255.255.255.0   U     0      0        0 enp0s3
```
If I enable nodeToNodeMesh or deploy the rr in the same subnet, the cross-subnet can work correctly.

Who can help me?

**Comment by fasaxc** [comment]
@nightmeng What route reflector are you using, are you using the Calico RR image?

**Comment by nightmeng** [comment]
@fasaxc Yes, I use the calico/routereflector:v0.3.1 image.

**Comment by robbrockbank** [comment]
Hi @nightmeng - yes it looks as though we'd need to update our bird templates to allow this - and may also require some additional bird integration work, so I don't have an immediate solution I'm afraid.

The issues is that we determine cross-subnet by using the bird `from` filter which corresponds to the router which the route has originated from - I assume with a RR that would be the address of the RR.

I haven't tried this out so it may not solve the issue for you: one thing you could try is using the `gobgp` routing backend - it is an option on the calico/node container configured through an environment variable.

Pass the following env into the calico/node container: `CALICO_NETWORKING_BACKEND=gobgp`
or if using `calicoctl node run` use the option `--backend=gobgp`

Let me know if that works as an interim solution.

**Comment by nightmeng** [comment]
I switch the calico backend from bird to gobgp, but it does not work for me. It seems to be another problem.

I deploy the calico/node on node-1 and node-2 and enable full mesh.
```
ETCD_ENDPOINTS=http://172.18.10.148:2379 calicoctl node run --node-image=172.16.200.19:5000/calico/node:v1.3.0 --ip=172.18.10.148 --backend=gobgp
ETCD_ENDPOINTS=http://172.18.10.148:2379 calicoctl node run --node-image=172.16.200.19:5000/calico/node:v1.3.0 --ip=172.18.10.143 --backend=gobgp
```
and then I get the following information:
node-1 (ip: 172.18.10.148):
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         172.18.10.254   0.0.0.0         UG    101    0        0 enp0s3
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
172.18.10.0     0.0.0.0         255.255.255.0   U     0      0        0 enp0s3
192.168.84.128  0.0.0.0         255.255.255.255 UH    0      0        0 cali3b8e77a9404
```
node-2 (ip: 172.18.10.143):
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         172.18.10.254   0.0.0.0         UG    101    0        0 enp0s3
10.0.3.0        172.18.10.148   255.255.255.0   UG    0      0        0 enp0s3
172.17.0.0      0.0.0.0         255.255.0.0     U     0      0        0 docker0
172.18.10.0     0.0.0.0         255.255.255.0   U     0      0        0 enp0s3
192.168.84.128  172.18.10.148   255.255.255.192 UG    0      0        0 enp0s3
```
The gobgp has add the route rule for node-2, but when I ping 192.168.84.128 from node-2 and then tcpdump from node-1:

```
[root@node-1 ~]# tcpdump -i any icmp -n
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked), capture size 65535 bytes
04:09:15.245424 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 1, length 64
04:09:16.245462 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 2, length 64
04:09:17.245388 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 3, length 64
04:09:18.245193 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 4, length 64
04:09:19.245178 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 5, length 64
04:09:20.245431 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 6, length 64
04:09:21.245044 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 7, length 64
04:09:22.245024 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 8, length 64
04:09:23.245269 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 9, length 64
04:09:24.245333 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 10, length 64
04:09:25.245397 IP 172.18.10.143 > 192.168.84.128: ICMP echo request, id 10840, seq 11, length 64
```
node-1 can receive the ICMP request from node-2, but it does reply the ICMP request.

**Comment by nightmeng** [comment]
@robbrockbank After I turned off nodeToNodeMesh and enabled route reflector, but disable ipip, I
can the bird.cfg from route reflector:

````

# ------------- RR as a node-specific peer -------------

# RR configured as a specific peer for calico node /host/node-1/peer_v4/172.18.12.121
protocol bgp Node_172_18_10_148 from bgp_template {
  local as 64512;
  neighbor 172.18.10.148 as 64512;
  rr client;
  rr cluster id 1.0.0.1;
}

# RR configured as a specific peer for calico node /host/node-2/peer_v4/172.18.12.121
protocol bgp Node_172_18_10_143 from bgp_template {
  local as 64512;
  neighbor 172.18.10.143 as 64512;
  rr client;
  rr cluster id 1.0.0.1;
}

```

And the route table also was updated:

node-1:
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         172.18.10.254   0.0.0.0         UG    101    0        0 enp0s3
172.18.10.0     0.0.0.0         255.255.255.0   U     0      0        0 enp0s3
192.168.84.128  0.0.0.0         255.255.255.255 UH    0      0        0 cali3b8e77a9404
```

node-2:
```
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         172.18.10.254   0.0.0.0         UG    101    0        0 enp0s3
172.18.10.0     0.0.0.0         255.255.255.0   U     0      0        0 enp0s3
192.168.84.128  172.18.10.148   255.255.255.192 UG    0      0        0 enp0s3
```

Then I enable ipip and specified ipip mode with "cross-subnet".

```
- apiVersion: v1
  kind: ipPool
  metadata:
    cidr: 192.168.0.0/16
  spec:
    ipip:
      enabled: true
      mode: cross-subnet
    nat-outgoing: true
- apiVersion: v1
  kind: ipPool
  metadata:
    cidr: fd80:24e2:f998:72d6::/64
  spec:
    nat-outgoing: true
```

node-1:

```
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         172.18.10.254   0.0.0.0         UG    101    0        0 enp0s3
172.18.10.0     0.0.0.0         255.255.255.0   U     0      0        0 enp0s3
192.168.84.128  0.0.0.0         255.255.255.255 UH    0      0        0 cali3b8e77a9404
192.168.247.0   172.18.10.143   255.255.255.192 UG    0      0        0 tunl0
```

node-2:

```
Kernel IP routing table
Destination     Gateway         Genmask         Flags Metric Ref    Use Iface
0.0.0.0         172.18.10.254   0.0.0.0         UG    101    0        0 enp0s3
172.18.10.0     0.0.0.0         255.255.255.0   U     0      0        0 enp0s3
192.168.84.128  172.18.10.148   255.255.255.192 UG    0      0        0 tunl0
```

It also switch the Iface from enp0s3 to tunl0  /(ㄒoㄒ)/~~

So the gobgp can not resolve this problem.

**Comment by nightmeng** [comment]
@robbrockbank hello, I modify the from filter to gw filter in bird_ipam.cfg.template of calico/node, and then it works for me, the following is the template file.

```
# Generated by confd
filter calico_pools {
  calico_aggr();
  custom_filters();
{{range ls "/v1/ipam/v4/pool"}}{{$data := json (getv (printf "/v1/ipam/v4/pool/%s" .))}}
  if ( net ~ {{$data.cidr}} ) then {
    accept;
  }
{{end}}
  reject;
}

{{$network_key := printf "/bgp/v1/host/%s/network_v4" (getenv "NODENAME")}}{{$network := getv $network_key}}
filter calico_ipip {
{{range ls "/v1/ipam/v4/pool"}}{{$data := json (getv (printf "/v1/ipam/v4/pool/%s" .))}}
  if ( net ~ {{$data.cidr}} ) then {
{{if $data.ipip_mode}}{{if eq $data.ipip_mode "cross-subnet"}}
    if ( gw ~ {{$network}} ) then
      krt_tunnel = "";                     {{/* Destination in ipPool, mode is cross sub-net, route from-host on subnet, do not use IPIP */}}
    else
      krt_tunnel = "{{$data.ipip}}";       {{/* Destination in ipPool, mode is cross sub-net, route from-host off subnet, set the tunnel (if IPIP not enabled, value will be "") */}}
    accept;
  } {{else}}
    krt_tunnel = "{{$data.ipip}}";         {{/* Destination in ipPool, mode not cross sub-net, set the tunnel (if IPIP not enabled, value will be "") */}}
    accept;
  } {{end}} {{else}}
    krt_tunnel = "{{$data.ipip}}";         {{/* Destination in ipPool, mode field is not present, set the tunnel (if IPIP not enabled, value will be "") */}}
    accept;
  } {{end}}
{{end}}
  accept;                                  {{/* Destination is not in any ipPool, accept  */}}
}
```

If I use this template file, does it bring up other new problems?

**Comment by nightmeng** [comment]
@caseydavenport I have tested the above fix method, and I want to trigger a pull request, but I can't find which project the bird_ipam.cfg.template is in, can you tell me?

**Comment by tmjd** [comment]
I believe the template you are looking for is in the project that this issue is on.  At `calico_node/filesystem/etc/calico/confd/templates`.  Here is a link to the folder it is in. https://github.com/projectcalico/calico/tree/master/calico_node/filesystem/etc/calico/confd/templates

**Comment by robbrockbank** [comment]
One thing to note here: it was originally the `gw` parameter we used for the cross-subnet check, but it didn't work in some circumstances (although I forget now which).  It's possible that the previous testing was flawed, but if we do go back to trying to use the gw parameter - we'd need to test thoroughly (esp. in AWS which was the primary use case for this feature).

**Comment by nightmeng** [comment]
@robbrockbank thank for your reply, I have tested this fix on my private cluster(with route reflector and without route reflector), and I do not tested it on any public cloud. That works for me now.

**Comment by nightmeng** [comment]
@robbrockbank I'm sorry for my mistake, we should use 'bgp_next_hop' parameter for cross-subnet check, the 'bgp_next_hop' can get the exactly IP address of next hop. 
The 'gw' parameter didn't work in some circumstances. when the next hop IP address  is from a different subnet, it will return the host gateway IP address.

**Comment by caseydavenport** [comment]
Fixes by #1104
